src/matlab2tikz.m

```diff
@@ -712,19 +714,24 @@ case guitypes()
         % A composite object might nest handleAllChildren calls that can
         % modify the m2t.currentHandleHasLegend value. Re-instate the
         % legend status. For detailed explanations check getLegendEntries().
-        m2t = hasLegendEntry(m2t,child);
-        [m2t,str] = addPlotyyReference(m2t, str, child);
-        str = addLegendInformation(m2t, str, child);
+        m2t               = hasLegendEntry(m2t,child);
+        [m2t, ref1, ref2] = addPlotyyReference(m2t, child);
+        legendInfo        = addLegendInformation(m2t, child);
+        % Add reference 2 BEFORE next plot to preserve color order
+        str               = join(m2t, {ref2, str, ref1, legendInfo}, '');
 
         % append the environment
-        pgfEnvironments{end+1} = str;
+        pgfEnvironments{envCounter} = str;
+        envCounter = envCounter +1;
     end
 end
 % ==============================================================================
-function [m2t, str] = addPlotyyReference(m2t, str, h)
+function [m2t, ref1, ref2] = addPlotyyReference(m2t, h)
```

Comment:
@miscco ref1 and ref2 are actually wrong if they refer to axis 1 and 2. Ref1 should is the latex label and ref2 should is the reference to that label.